src/v/cloud_topics/level_one/offset_interval_set.cc

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cloud_topics/level_one/offset_interval_set.h"
+
+#include "model/fundamental.h"
+
+namespace experimental::cloud_topics::l1 {
+
+bool offset_interval_set::stream::has_next() const noexcept {
+    return iter_ != set_.end();
+}
+
+offset_interval_set::interval offset_interval_set::stream::next() {
+    vassert(has_next(), "next() called while has_next() is false");
+    interval ret{
+      .base_offset = kafka::offset(iter_->first),
+      .last_offset = kafka::offset(iter_->second - 1),
```

Comment:
Another approach is to use boost::transform_iterators or use an as_range function that returns the iset as a transformed range using https://en.cppreference.com/w/cpp/ranges/transform_view.html

Just mentioning as alternatives, no real problem with this iterator anyways.